Quando trabalhamos com sequências de dados que representam pontos no tempo (time series) é muito comum entender o crescimento de um valor em função do tempo. Por exemplo, o preço de um produto ou uma ação que antes era 100 e agora 110 cresceu 110/100 - 1 = 10%, e depois se caiu para 105, caiu 105/110 - 1 = -4%.

Portanto a fórmula geral é valor / valor_anterior - 1. Em python podemos gerar alguns dados aleatórios:

In [1]:
import numpy as np
import pandas as pd

np.random.seed(234)
aleatorios = np.random.randint(low = 950, high = 1050, size=100)
df = pd.DataFrame(aleatorios, columns=["preco"])
df.head()

preco
0   1022
1    981
2   1018
3   1007
4   1045

Como o valor da linha n+1 vai ser dividido pelo valor da linha n temos algo como:

In [2]:
# linhas na posicao posterior (n+1): df.preco.shift(-1)
# linhas na posicao atual (n): df.preco

df['crescimento'] = df.preco.shift(-1).div(df.preco) - 1
df.head()

preco  crescimento
0   1022    -0.040117
1    981     0.037717
2   1018    -0.010806
3   1007     0.037736
4   1045    -0.059330

1022-981 = 41
41/1022 = -0.040117

Mas, claro, como toda boa análise básica de time series, o pandas já fornece uma função do gênero para trabalharmos:

In [5]:
df['crescimento_direto'] = df.preco.pct_change(periods=-1)
df.head()

preco  crescimento  crescimento_direto
0   1022    -0.040117            0.041794
1    981     0.037717           -0.036346
2   1018    -0.010806            0.010924
3   1007     0.037736           -0.036364
4   1045    -0.059330            0.063072